WTG.Analyzers/Analyzers/DiscardLocal/DiscardLocalAnalyzer.cs

```diff
@@ -0,0 +1,106 @@
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using WTG.Analyzers.Utils;
+
+namespace WTG.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public sealed partial class DiscardLocalAnalyzer : DiagnosticAnalyzer
+	{
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+			Rules.LocalCouldBeConfusedWithDiscardRule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction(CompilationStart);
+		}
+
+		static void CompilationStart(CompilationStartAnalysisContext context)
+		{
+			var cache = new FileDetailCache();
+
+			context.RegisterSyntaxNodeAction(
+				c => Analyze(c, cache),
+				SyntaxKind.Argument,
+				SyntaxKind.ForEachStatement,
+				SyntaxKind.Parameter,
+				SyntaxKind.VariableDeclarator);
+		}
+
+		static void Analyze(SyntaxNodeAnalysisContext context, FileDetailCache cache)
+		{
+			if (cache.IsGenerated(context.SemanticModel.SyntaxTree, context.CancellationToken))
+			{
+				return;
+			}
+
+			switch (context.Node.Kind())
+			{
+				case SyntaxKind.Argument:
+					Analyze(context, (ArgumentSyntax)context.Node);
+					return;
+
+				case SyntaxKind.ForEachStatement:
+					Analyze(context, (ForEachStatementSyntax)context.Node);
+					return;
+
+				case SyntaxKind.Parameter:
+					Analyze(context, (ParameterSyntax)context.Node);
+					return;
+
+				case SyntaxKind.VariableDeclarator:
+					Analyze(context, (VariableDeclaratorSyntax)context.Node);
+					return;
+			}
+		}
+
+		static void Analyze(SyntaxNodeAnalysisContext context, ArgumentSyntax node)
+		{
+			if (node.RefOrOutKeyword.IsKind(SyntaxKind.OutKeyword) &&
+				node.Expression.IsKind(SyntaxKind.DeclarationExpression))
+			{
+				var declaration = (DeclarationExpressionSyntax)node.Expression;
+
+				if (declaration.Designation.IsKind(SyntaxKind.DiscardDesignation))
```

Comment:
This is interesting. It gets analyzed as a Discard, even though its not?

Reply:
yeah, I thought that was odd too.